designs/2024-support-ts-config-files/README.md

```diff
@@ -0,0 +1,341 @@
+
+```ts
+import eslint from "@eslint/js"
+import type { Linter } from "eslint"
+
+const config: Linter.FlatConfig[] = [
+  eslint.configs.recommended,
+  {
+    rules: {
+      "no-console": [0],
+    },
+  },
+]
+
+export default config
+```
+
+It is worth noting that you can already perform some type-checking with the [`checkJs`](https://www.typescriptlang.org/tsconfig#checkJs) TypeScript option and a JavaScript config file. Here's an example:
+
+with `eslint.config.mjs` or (`eslint.config.js` + `"type": "module"` in the nearest `package.json`):
+
+```js
+import eslint from "@eslint/js"
+
+/** @type {import('eslint').Linter.FlatConfig[]} */
+const config = [
+  eslint.configs.recommended,
+  {
+    rules: {
+      "no-console": [0],
+    },
+  },
+]
+
+export default config
+```
+
+with `eslint.config.cjs` or (`eslint.config.js` without `"type": "module"` in the nearest `package.json`):
+
+```js
+const eslint = require("@eslint/js")
+
+/** @type {import('eslint').Linter.FlatConfig[]} */
+const config = [
+  eslint.configs.recommended,
+  {
+    rules: {
+      "no-console": [0],
+    },
+  },
+]
+
+module.exports = config
+```
+
+## Documentation
+
+<!--
+    How will this RFC be documented? Does it need a formal announcement
+    on the ESLint blog to explain the motivation?
+-->
+
+The documentation for this feature will be added to the [ESLint User Guide](https://eslint.org/docs/user-guide/configuring) page. The documentation will explain how to use TypeScript configuration files and the differences between JavaScript and TypeScript configuration files.
+
+## Drawbacks
+
+<!--
+    Why should we *not* do this? Consider why adding this into ESLint
+    might not benefit the project or the community. Attempt to think
+    about any opposing viewpoints that reviewers might bring up.
+
+    Any change has potential downsides, including increased maintenance
+    burden, incompatibility with other tools, breaking existing user
+    experience, etc. Try to identify as many potential problems with
+    implementing this RFC as possible.
+-->
+
+This change will most likely require at least one external tool as either a dependency or a peer dependency.
+
+## Backwards Compatibility Analysis
+
+<!--
+    How does this change affect existing ESLint users? Will any behavior
+    change for them? If so, how are you going to minimize the disruption
+    to existing users?
+-->
+
+This goal is to minimize the disruption to existing users. The primary focus is to ensure that the existing JavaScript configuration files continue to work as expected. The changes will only affect TypeScript users who are using TypeScript configuration files. The changes will not affect the existing JavaScript configuration files.
+
+## Alternatives
+
+<!--
+    What other designs did you consider? Why did you decide against those?
+
+    This section should also include prior art, such as whether similar
+    projects have already implemented a similar feature.
+-->
+
```

Comment:
there is a community project: https://github.com/antfu/eslint-ts-patch to support it.  would like to hear the author :) / @antfu

Reply:
It uses [jiti](https://github.com/unjs/jiti) to support ts file. Which already mentioned https://github.com/eslint/rfcs/pull/117#issuecomment-2036336411 that it doesn't currently support top-level await.

I would personally recommend using https://github.com/egoist/bundle-require instead which is more robust and will respect tsconfig.json. The downside is that it would introduce `esbuild` into the dependency. If the install size is a concern, I guess we could have an optional package like `@eslint/config-loader-ts` that only requires when users use ts version of config.

Reply:
Yes, the inspector uses that because we need to know the dependencies of the config to do automatic reloads. Supporting TS was a free side-effect. 

Even if ESLint doesn't need information on dependencies, I think it's still a good way to support TS. Vite uses the same approach to load `vite.config.ts`.

Reply:
`bundle-require` writes a temp file besides the original right? These temp files are a major source of problems with vite, see https://github.com/vitejs/vite/issues/9470.

Reply:
@antfu 

> Does bundle-require write a temp file to disk like vite does?

Reply:
Yes it does. But I think currently there is no better solution unless Node supports TS natively, or we pass the loader flag in our CLI: https://tsx.is/node#es-modules-only

Reply:
So it sounds like the deciding factor here is whether or not we think that top-level await is important to support in TypeScript config files. If it *is* important, then `bundle-require` is the only option with the caveat that it creates a temp file; if it *isn't* important, then we can use Jiti.

What are people's opinions on whether or not top-level await is important?

Reply:
Can't think of any case where I wold use TLA in any config file. If I need to do file I/O, I would do is synchronous because that is faster than the async variants. If a user would want to `fetch` it would likely be unavoidable to have some support for async.

You could also avoid TLA by accepting a function that return either a promise or sync result, e.g.

```js
export default async () => {files: []};
export default () => {files: []};

Reply:
I agree with @silverwind here, I think [`jiti`](https://www.npmjs.com/package/jiti) would be more appropriate in this case since top-level `await` expressions are not something I see people using in an ESLint config file. [`jiti`](https://www.npmjs.com/package/jiti) also has the benfit of ignoring `tsconfig.json` files which according to [this comment](https://github.com/eslint/rfcs/pull/117#discussion_r1543602282) is the desired outcome here. It also doesn't write temp files to disk which as mentioned in [this comment](https://github.com/eslint/rfcs/pull/117#discussion_r1561318636) can potentially cause some issues.

Reply:
tsx v4.9 exports `tsImport()` which supports loading TypeScript files with TLA (compiles in-memory).

https://tsx.is/node#enhanced-import-require

Reply:
@privatenumber That's wonderful news I'll give it a shot.

Reply:
I agree that TLA shouldn't be a blocker for this. :+1:

Reply:
tsx is mentioned as a non-viable alternative because it uses custom loaders. Does that also hold when `tsImport()` is used?

Reply:
Personally, I see TLA as a language spec requirement, and I don't think whether to support it should be up to whether we can imagine use-cases.

But if we want evaluate usage, here are examples of TLA in configs (Vite, Rollup, Next, Jest, esbuild, etc):
https://github.com/search?q=path%3A%2F%5C.config%5C.%28js%7Cts%29%24%2F+%2F%5Eawait+%2F&type=code


----

> tsx is mentioned as a non-viable alternative because it uses custom loaders. Does that also hold when `tsImport()` is used?

`tsImport()` is isolated and won't add TypeScript support to the rest of the environment.

But just to be clear, I'm only sharing it as an option. Personally, I would like to have popular compilers supported as optional peer dependencies to accommodate different preferences and environments.

Reply:
> Personally, I see TLA as a language spec requirement, and I don't think whether to support it should be up to whether we can imagine use-cases.
> 
> But if we want evaluate usage, here are examples of TLA in configs (Vite, Rollup, Next, Jest, esbuild, etc): https://github.com/search?q=path%3A%2F%5C.config%5C.%28js%7Cts%29%24%2F+%2F%5Eawait+%2F&type=code

Interesting. There are also ESLint configs that use TLA: https://github.com/search?q=path%3A%2Feslint%5C.config%5C.js%24%2F+%2F%5Econst+.%2B+%3D+await+%2F&type=code.

Reply:
Jiti won't recognize TLA not only in the config file itself, but also in imported modules, because they are all treated as CommonJS. I did a quick test using a patched version of ESLint with the changes from [#18134](https://github.com/eslint/eslint/pull/18134) and with this config:

```ts
// eslint.config.ts
export { default } from './recommended.mjs';
```

```js
// recommended.mjs
import { readFile } from 'node:fs/promises';

const json = await readFile('package.json', 'utf-8');
const { name } = JSON.parse(json);

export default [{ name, rules: { 'no-undef': 'error' } }];
```

When running `eslint`, I got an error as expected:

```
Oops! Something went wrong! :(

ESLint: 9.2.0

ReferenceError: await is not defined
    at ../project/recommended.mjs:4:14
    at evalModule (../project/node_modules/eslint/node_modules/jiti/dist/jiti.js:1:256443)
    at jiti (../project/node_modules/eslint/node_modules/jiti/dist/jiti.js:1:254371)
    at ../project/eslint.config.ts:2:43
    at evalModule (../project/node_modules/eslint/node_modules/jiti/dist/jiti.js:1:256443)
    at jiti (../project/node_modules/eslint/node_modules/jiti/dist/jiti.js:1:254371)
    at loadFlatConfigFile (../project/node_modules/eslint/lib/eslint/eslint.js:335:24)
    at async calculateConfigArray (../project/node_modules/eslint/lib/eslint/eslint.js:421:28)
    at async ESLint.lintFiles (../project/node_modules/eslint/lib/eslint/eslint.js:840:25)
    at async Object.execute (../project/node_modules/eslint/lib/cli.js:500:23)
    at async main (../project/node_modules/eslint/bin/eslint.js:153:22)
```

The workaround of using a dynamic import didn't help either, and resulted in the same error:

```ts
// eslint.config.ts
export default (async () =>
    (await import('./recommended.mjs')).default
)();
```

This means that if we decide to use Jiti, plugin developers should be advised to avoid TLA in their shared configs (including rules and transitive dependencies), or else those configs will not work for users who have a `eslint.config.ts` in their project.

Reply:
I added support for all the ts loaders we mentioned in [`eslint-ts-patch`](https://github.com/antfu/eslint-ts-patch#typescript-loaders) and listed their trade-offs, where you can give them a try today.

Reply:
@privatenumber I tried to use tsx `tsImport` as explained [here](https://github.com/privatenumber/tsx/blob/master/docs/node.md) to load a `.ts` config file, but didn't succeed. This is my code:

```js
let config;
const { tsImport, register } = require("tsx/esm/api");
const unregister = register();
try {
    config = (await tsImport(fileURL.href, __filename)).default;
} finally {
    unregister();
}
```

where `fileURL` is the URL of this config file:

```ts
// eslint.config.mts
export default [];
```

The error stack trace indicates that the config file is being loaded as CommonJS:

```
SyntaxError: Unexpected token 'export'
    at wrapSafe (node:internal/modules/cjs/loader:1389:18)
    at Module._compile (node:internal/modules/cjs/loader:1425:20)
    at Module._extensions..js (node:internal/modules/cjs/loader:1564:10)
    at Module.load (node:internal/modules/cjs/loader:1287:32)
    at Module._load (node:internal/modules/cjs/loader:1103:12)
    at cjsLoader (node:internal/modules/esm/translators:318:15)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:258:7)
    at ModuleJob.run (node:internal/modules/esm/module_job:262:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:474:24)
```

I guess I'm doing something wrong?

Reply:
This is probably not the best place to discuss this. Can you send me a reproduction link in the tsx repo?

Reply:
> This is probably not the best place to discuss this. Can you send me a reproduction link in the tsx repo?

Thanks! I think the problem is that I wasn't calling `register` from `require("tsx/cjs/api")`. It works if I add that call before `tsImport`. Another problem is that there is seemingly no way to completely undo the changes done by `tsImport` to the Node.js loader internals. Particularly, I think there is no way to unregister the [ESM `Module` loader](https://github.com/privatenumber/tsx/blob/4f515ab0bb8f3c8c610dc57556b978cbf90a5560/src/esm/api/register.ts#L22-L33). I didn't even find out how to do that programmatically, so this is possibly a limitation of Node.js.